File: hdf-cli/pkg/diff/matching/fuzzy_match.go
package matching

import (
	"regexp"
	"sort"
	"strings"

	hdf "github.com/mitre/hdf-cli/pkg/hdf"
)

// defaultMinConfidence is the default minimum Jaccard similarity threshold.
const defaultMinConfidence = 0.6

// stopWords contains common English stop words to exclude from tokenization.
var stopWords = map[string]bool{
	"a": true, "an": true, "the": true, "is": true, "are": true,
	"was": true, "were": true, "be": true, "been": true, "being": true,
	"have": true, "has": true, "had": true, "do": true, "does": true,
	"did": true, "will": true, "would": true, "could": true,
	"should": true, "may": true, "might": true, "shall": true, "can": true,
	"to": true, "of": true, "in": true, "for": true,
	"on": true, "with": true, "at": true, "by": true, "from": true,
	"as": true, "into": true, "through": true, "during": true,
	"before": true, "after": true, "above": true, "below": true,
	"between": true, "out": true, "off": true, "over": true,
	"under": true, "again": true, "further": true, "then": true,
	"once": true, "and": true, "but": true, "or": true, "nor": true,
	"not": true, "no": true, "so": true, "if": true, "it": true,
	"its": true, "that": true, "this": true, "these": true, "those": true,
	"each": true, "every": true, "all": true, "both": true, "few": true,
	"more": true, "most": true, "other": true, "some": true,
	"such": true, "than": true, "too": true, "very": true, "just": true,
	"about": true,
}

// splitRegex splits on non-alphanumeric characters.
var splitRegex = regexp.MustCompile(`[^a-z0-9]+`)

// Tokenize splits a title string into a set of meaningful tokens.
// It lowercases the string, splits on non-alphanumeric characters,
// filters out stop words and tokens with length <= 2, and returns unique tokens.
func Tokenize(text string) []string {
	if text == "" {
		return nil
	}

	lower := strings.ToLower(text)
	parts := splitRegex.Split(lower, -1)

	seen := make(map[string]bool)
	var result []string
	for _, token := range parts {
		if len(token) <= 2 {
			continue
		}
		if stopWords[token] {
			continue
		}
		if !seen[token] {
			seen[token] = true
			result = append(result, token)
		}
	}
	return result
}

// JaccardSimilarity computes the Jaccard similarity between two sets.
// Returns |intersection| / |union|, or 0 if both sets are empty.
func JaccardSimilarity(a, b map[string]bool) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 0.0
	}

	// Iterate over the smaller set for efficiency
	smaller, larger := a, b
	if len(a) > len(b) {
		smaller, larger = b, a
	}

	intersectionSize := 0
	for item := range smaller {
		if larger[item] {
			intersectionSize++
		}
	}

	unionSize := len(a) + len(b) - intersectionSize
	if unionSize == 0 {
		return 0.0
	}

	return float64(intersectionSize) / float64(unionSize)
}

// tokenSet converts a slice of tokens to a set (map[string]bool).
func tokenSet(tokens []string) map[string]bool {
	s := make(map[string]bool, len(tokens))
	for _, t := range tokens {
		s[t] = true
	}
	return s
}

// FuzzyTitleStrategy matches requirements by token-based Jaccard similarity on titles.
type FuzzyTitleStrategy struct {
	threshold float64
}

// NewFuzzyTitleStrategy creates a new fuzzy title matching strategy.
// If minConfidence is 0 or negative, the default threshold of 0.6 is used.
func NewFuzzyTitleStrategy(minConfidence float64) *FuzzyTitleStrategy {
	if minConfidence <= 0 {
		minConfidence = defaultMinConfidence
	}
	return &FuzzyTitleStrategy{threshold: minConfidence}
}

// Name returns the strategy name.
func (s *FuzzyTitleStrategy) Name() string {
	return "fuzzyTitle"
}

type candidate struct {
	oldIdx     int
	newIdx     int
	similarity float64
}

// Match matches requirements by fuzzy title similarity using greedy best-match.
func (s *FuzzyTitleStrategy) Match(oldReqs, newReqs []hdf.EvaluatedRequirement) MatchResult {
	result := MatchResult{}

	// Pre-tokenize all titles
	type indexedTokens struct {
		idx    int
		tokens map[string]bool
	}

	var oldTokens []indexedTokens
	for i, req := range oldReqs {
		if req.Title != nil && *req.Title != "" {
			tokens := Tokenize(*req.Title)
			if len(tokens) > 0 {
				oldTokens = append(oldTokens, indexedTokens{idx: i, tokens: tokenSet(tokens)})
			}
		}
	}

	var newTokens []indexedTokens
	for i, req := range newReqs {
		if req.Title != nil && *req.Title != "" {
			tokens := Tokenize(*req.Title)
			if len(tokens) > 0 {
				newTokens = append(newTokens, indexedTokens{idx: i, tokens: tokenSet(tokens)})
			}
		}
	}

	// Track matched indices
	matchedOldIndices := make(map[int]bool)
	matchedNewIndices := make(map[int]bool)

	// Build all potential matches with similarity scores
	var candidates []candidate
	for _, old := range oldTokens {
		for _, nw := range newTokens {
			sim := JaccardSimilarity(old.tokens, nw.tokens)
			if sim >= s.threshold {
				candidates = append(candidates, candidate{
					oldIdx:     old.idx,
					newIdx:     nw.idx,
					similarity: sim,
				})
			}
		}
	}

	// Sort by similarity descending (greedy best-match)
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i].similarity > candidates[j].similarity
	})

	// Greedily assign matches
	for _, c := range candidates {
		if matchedOldIndices[c.oldIdx] || matchedNewIndices[c.newIdx] {
			continue
		}

		result.Matched = append(result.Matched, MatchPair{
			OldReq:     oldReqs[c.oldIdx],
			NewReq:     newReqs[c.newIdx],
			Strategy:   "fuzzyTitle",
			Confidence: c.similarity,
		})
		matchedOldIndices[c.oldIdx] = true
		matchedNewIndices[c.newIdx] = true
	}

	// Collect unmatched
	for i, req := range oldReqs {
		if !matchedOldIndices[i] {
			result.UnmatchedOld = append(result.UnmatchedOld, req)
		}
	}
	for i, req := range newReqs {
		if !matchedNewIndices[i] {
			result.UnmatchedNew = append(result.UnmatchedNew, req)
		}
	}

	return result
}

File: hdf-diff/src/matching/fuzzy-match.ts
import type { MatchStrategy, MatchResult } from './types.js';

/** Common English stop words to exclude from tokenization */
const STOP_WORDS = new Set([
  'a', 'an', 'the', 'is', 'are', 'was', 'were', 'be', 'been', 'being',
  'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could',
  'should', 'may', 'might', 'shall', 'can', 'to', 'of', 'in', 'for',
  'on', 'with', 'at', 'by', 'from', 'as', 'into', 'through', 'during',
  'before', 'after', 'above', 'below', 'between', 'out', 'off', 'over',
  'under', 'again', 'further', 'then', 'once', 'and', 'but', 'or', 'nor',
  'not', 'no', 'so', 'if', 'it', 'its', 'that', 'this', 'these', 'those',
  'each', 'every', 'all', 'both', 'few', 'more', 'most', 'other', 'some',
  'such', 'than', 'too', 'very', 'just', 'about',
]);

/**
 * Tokenize a title string into a set of meaningful tokens.
 *
 * - Lowercases the string
 * - Splits on whitespace and punctuation
 * - Filters out common stop words
 * - Returns unique tokens as a Set
 */
export function tokenize(text: string): Set<string> {
  if (!text) return new Set();

  const tokens = text
    .toLowerCase()
    .split(/[\s\-_/\\,.;:!?()[\]{}"']+/)
    .filter((t) => t.length > 0 && !STOP_WORDS.has(t));

  return new Set(tokens);
}

/**
 * Compute Jaccard similarity between two sets.
 * Returns |intersection| / |union|, or 0 if both sets are empty.
 */
export function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0.0;

  let intersectionSize = 0;
  // Iterate over the smaller set for efficiency
  const [smaller, larger] = a.size <= b.size ? [a, b] : [b, a];
  for (const item of smaller) {
    if (larger.has(item)) {
      intersectionSize++;
    }
  }

  const unionSize = a.size + b.size - intersectionSize;
  // Unreachable: if both sets are empty, we return above. If either is non-empty, unionSize > 0.
  // Guard kept for mathematical correctness.
  /* c8 ignore next */
  if (unionSize === 0) return 0.0;

  return intersectionSize / unionSize;
}

/**
 * Default minimum confidence threshold for fuzzy title matching.
 */
const DEFAULT_MIN_CONFIDENCE = 0.6;

/**
 * Create a fuzzy title matching strategy.
 *
 * Matches requirements by token-based Jaccard similarity on the `title` field.
 * Uses greedy best-match: for each unmatched old requirement, finds the
 * best-matching unmatched new requirement above the confidence threshold.
 *
 * @param minConfidence Minimum Jaccard similarity to accept a match (default: 0.6)
 */
export function createFuzzyTitleStrategy(minConfidence?: number): MatchStrategy {
  const threshold = minConfidence ?? DEFAULT_MIN_CONFIDENCE;

  return {
    name: 'fuzzyTitle',
    match(oldReqs: Record<string, unknown>[], newReqs: Record<string, unknown>[]): MatchResult {
      const result: MatchResult = {
        matched: [],
        unmatchedOld: [],
        unmatchedNew: [],
      };

      // Pre-tokenize all titles
      const oldTokens: Array<{ idx: number; tokens: Set<string> }> = [];
      for (let i = 0; i < oldReqs.length; i++) {
        const title = oldReqs[i]!['title'];
        if (typeof title === 'string' && title.length > 0) {
          const tokens = tokenize(title);
          if (tokens.size > 0) {
            oldTokens.push({ idx: i, tokens });
          }
        }
      }

      const newTokens: Array<{ idx: number; tokens: Set<string> }> = [];
      for (let i = 0; i < newReqs.length; i++) {
        const title = newReqs[i]!['title'];
        if (typeof title === 'string' && title.length > 0) {
          const tokens = tokenize(title);
          if (tokens.size > 0) {
            newTokens.push({ idx: i, tokens });
          }
        }
      }

      // Track matched indices
      const matchedOldIndices = new Set<number>();
      const matchedNewIndices = new Set<number>();

      // Build a list of all potential matches with similarity scores
      const candidates: Array<{
        oldIdx: number;
        newIdx: number;
        similarity: number;
      }> = [];

      for (const old of oldTokens) {
        for (const nw of newTokens) {
          const sim = jaccardSimilarity(old.tokens, nw.tokens);
          if (sim >= threshold) {
            candidates.push({
              oldIdx: old.idx,
              newIdx: nw.idx,
              similarity: sim,
            });
          }
        }
      }

      // Sort by similarity descending (greedy best-match)
      candidates.sort((a, b) => b.similarity - a.similarity);

      // Greedily assign matches
      for (const candidate of candidates) {
        if (matchedOldIndices.has(candidate.oldIdx) || matchedNewIndices.has(candidate.newIdx)) {
          continue;
        }

        result.matched.push({
          oldReq: oldReqs[candidate.oldIdx]!,
          newReq: newReqs[candidate.newIdx]!,
          strategy: 'fuzzyTitle',
          confidence: candidate.similarity,
        });
        matchedOldIndices.add(candidate.oldIdx);
        matchedNewIndices.add(candidate.newIdx);
      }

      // Collect unmatched
      for (let i = 0; i < oldReqs.length; i++) {
        if (!matchedOldIndices.has(i)) {
          result.unmatchedOld.push(oldReqs[i]!);
        }
      }
      for (let i = 0; i < newReqs.length; i++) {
        if (!matchedNewIndices.has(i)) {
          result.unmatchedNew.push(newReqs[i]!);
        }
      }

      return result;
    },
  };
}
